### PR TITLE
[pytorch][CI] end-to-end custom build script

### PR DESCRIPTION
**Pull request by ljk53**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #34055 [pytorch][mobile] support for custom mobile build with dynamic dispatch
* #34038 [pytorch][mobile] make sure mobile build work with dynamic dispatch
* **#34012 [pytorch][CI] end-to-end custom build script**

Summary:
Today some mobile simulator tests only run on landed PRs and it requires
setting up special build environment to repro errors locally.

The goal of the PR is to do end-to-end mobile custom build & integration
tests with host toolchain (using same CMake options as mobile build). This
way, non-mobile engineers can capture & debug mobile related build issues
much more easily.

There are three custom build types that this script supports:

1. `TEST_DEFAULT_BUILD=1 ./build.sh` - it is similar to the prebuilt libtorch
libraries released for Android and iOS (same CMake build options + host
toolchain), which doesn't contain autograd function nor backward ops thus is
smaller than full LibTorch.

2. `TEST_CUSTOM_BUILD_STATIC=1 ./build.sh` - it further optimizes libtorch
size by only including ops used by a specific model.

3. `TEST_CUSTOM_BUILD_DYNAMIC=1 ./build.sh` - similar as 2) except that it
relies on the op dependency graph (instead of static dispatch) to calculate
and keep all transitively dependent ops by the model.

Type 2) will be deprecated by type 3) in the future.
Type 3) custom build has not been fully supported yet so it's expected to fail.

Replacing existing mobile build CI to run Type 1) build & integration test.

Differential Revision: [D20193328](https://our.internmc.facebook.com/intern/diff/D20193328)

### CODE REVIEW

File: .jenkins/pytorch/build-mobile.sh [hunk 1]
```diff
@@ -10,12 +10,13 @@ COMPACT_JOB_NAME="${BUILD_ENVIRONMENT}"
 
 source "$(dirname "${BASH_SOURCE[0]}")/common.sh"
 
-echo "Clang version:"
-clang --version
+# Install torch & torchvision - used to download & trace test model.
+pip install torch torchvision --progress-bar off
```
Comment (ljk53) [inline .jenkins/pytorch/build-mobile.sh]: I see iOS simulator test is doing this - I'm not entirely sure whether it's a good idea to download and install torch & torchvision on every PR.

My goal is to download & trace & run a non-trivial model so I chose MobileNetV2 - shall I create a dummy TorchScript model with similar set of ops like MobileNetV2 but with small dummy weights instead so that I can commit the test model without downloading it from network?
Are there any similar integration tests for non-mobile builds that I can reuse?

File: test/mobile/custom_build/build.sh [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+  cd "${BUILD_ROOT}"
+  MODEL="${BUILD_ROOT}/MobileNetV2.pt"
+  ROOT_OPS="${BUILD_ROOT}/MobileNetV2.yaml"
+
+  if [ ! -f "${MODEL}" ] || [ ! -f "${ROOT_OPS}" ]; then
+    python "${TEST_SRC_ROOT}/prepare_model.py"
+  fi
+}
+
+generate_op_dependency_graph() {
```
Comment (ljk53) [inline test/mobile/custom_build/build.sh]: selected mobile build with dynamic dispatch is not supported in OSS yet so this won't be called now.

When I enable it in the future, it will build libtorch twice - first generate the op dependency graph inline (which will build mobile libtorch into LLVM bitcode), then call the regular mobile build (which will build mobile libtorch again).

I could reuse the output of the other code-analysis CI's output to save some work, but it won't reduce latency of the job. Fortunately it took much less time to build libtorch with mobile CMake options (skipped autograd and caffe2) so it's probably not going to be too bad.

Let me know if you have better suggestions.

File: .jenkins/pytorch/build-mobile.sh (hunk 1 above)
Comment (ezyang) [reply .jenkins/pytorch/build-mobile.sh]: It's not *great* (I commented on this at https://github.com/pytorch/pytorch/pull/30133#issuecomment-557130164) but it's certainly the most convenient way to dump the model (and we do this in a few other places). An alternative is to save the model in s3 directly and then suck it down directly, but that makes it more difficult to update the model later.

Comment (ezyang) [reply .jenkins/pytorch/build-mobile.sh]: As for your other question, let me read the patch and get back to you on it.

File: test/mobile/custom_build/prepare_model.py [hunk 3]
```diff
@@ -0,0 +1,17 @@
+import torch
+import torchvision
+import yaml
+
+# Download and trace the model.
+model = torchvision.models.mobilenet_v2(pretrained=True)
+model.eval()
+example = torch.rand(1, 3, 224, 224)
+traced_script_module = torch.jit.trace(model, example)
```
Comment (ezyang) [inline test/mobile/custom_build/prepare_model.py]: Unrelated question: how come we're tracing here (and not using TorchScript, e.g.?)

File: test/mobile/custom_build/predictor.cpp [hunk 4]
```diff
@@ -0,0 +1,52 @@
+#include <iostream>
```
Comment (ezyang) [inline test/mobile/custom_build/predictor.cpp]: Would be nice to have a little explanation of what this does.

File: test/mobile/custom_build/predictor.cpp [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+using namespace std;
+
+namespace {
+
+struct MobileCallGuard {
+  // AutoGrad is disabled for mobile by default.
+  torch::autograd::AutoGradMode no_autograd_guard{false};
+  // Disable graph optimizer to ensure list of unused ops are not changed for
+  // custom mobile build.
+  torch::jit::GraphOptimizerEnabledGuard no_optimizer_guard{false};
```
Comment (ezyang) [inline test/mobile/custom_build/predictor.cpp]: The reasoning here makes sense, but in general, this seems a bit questionable, since I imagine that one might quite reasonably want to optimize graphs for mobile build. Wouldn't it be "more correct" to optimize first, and then compute the list of unused ops? Can this be conveniently supported?

I understand the goal of this PR is to get an e2e CI running, so this wouldn't be blocking, but it's something to think about...

File: test/mobile/custom_build/predictor.cpp [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+};
+
+void init() {
+  auto qengines = at::globalContext().supportedQEngines();
+  if (std::find(qengines.begin(), qengines.end(), at::QEngine::QNNPACK) !=
+      qengines.end()) {
+    at::globalContext().setQEngine(at::QEngine::QNNPACK);
```
Comment (ezyang) [inline test/mobile/custom_build/predictor.cpp]: Also not really related to this PR, but is there really any reason why you wouldn't want the lack of QNNPACK to be a hard error? This code (1) seems annoying boilerplate that everyone has to write, and (2) seems like it would mask build problems when QNNPACK is not available.

File: test/mobile/custom_build/build.sh [hunk 7]
```diff
@@ -0,0 +1,145 @@
+#!/bin/bash
+###############################################################################
+# This script shows/tests the flow to build libtorch locally with optimized
```
Comment (ezyang) [inline test/mobile/custom_build/build.sh]: Mentioning a pet peeve of mine: I generally don't like build scripts intended for general use by users, because they can make it difficult to toggle build parameters on the underlying cmake invocations (we had this problem for the longest time in setup.py until xuhdev wrote a general framework for passing environment variables into cmake flags for the underlying cmake invocation).

File: test/mobile/custom_build/build.sh [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+# smaller than full LibTorch.
+#
+# 2. `TEST_CUSTOM_BUILD_STATIC=1 ./build.sh` - it further optimizes libtorch
+# size by only including ops used by a specific model.
+#
+# 3. `TEST_CUSTOM_BUILD_DYNAMIC=1 ./build.sh` - similar as 2) except that it
+# relies on the op dependency graph (instead of static dispatch) to calculate
+# and keep all transitively dependent ops by the model.
+# Note that LLVM_DIR environment variable should be set to the location of
+# LLVM-dev toolchain.
```
Comment (ezyang) [inline test/mobile/custom_build/build.sh]: Is this intended to be an example for users for how they set things up? :)

File: test/mobile/custom_build/build.sh [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+mkdir -p "${BUILD_ROOT}"
+cd "${BUILD_ROOT}"
+
+prepare_model_and_dump_root_ops() {
+  cd "${BUILD_ROOT}"
+  MODEL="${BUILD_ROOT}/MobileNetV2.pt"
+  ROOT_OPS="${BUILD_ROOT}/MobileNetV2.yaml"
+
+  if [ ! -f "${MODEL}" ] || [ ! -f "${ROOT_OPS}" ]; then
+    python "${TEST_SRC_ROOT}/prepare_model.py"
```
Comment (ezyang) [inline test/mobile/custom_build/build.sh]: If this script is intended to be run locally by developers who might run this repeatedly, it might be a good idea to unconditionally rerun `prepare_model.py` if its mtime is greater than MODEL/ROOT_OPS

File: test/mobile/custom_build/build.sh (hunk 2 above)
Comment (ezyang) [reply test/mobile/custom_build/build.sh]: This is the terminal state of the static analyzer, is that right? Because we cannot use LLVM to generate the op dependency graph without compiling everything, and then we need to build the actual build. Do you have some sense of how long these steps take?

It's kind of irritating that you have to build twice, but assuming you need LLVM bitcode I'm not sure how to get around it. I suppose that for OSS end users, we could directly distribute LLVM bitcode for libtorch, so they don't need to do this step.

File: test/mobile/custom_build/build.sh (hunk 9 above)
Comment (ezyang) [reply test/mobile/custom_build/build.sh]: Kind of wondering if this should have been written in make/cmake :)

**Review by ezyang — APPROVED** [review]
OK, looks reasonable.

File: test/mobile/custom_build/prepare_model.py (hunk 3 above)
Comment (ljk53) [reply test/mobile/custom_build/prepare_model.py]: It's copied from a very old example, so the latest approach should be `script_model = torch.jit.script(mode)`?

File: .jenkins/pytorch/build-mobile.sh (hunk 1 above)
Comment (ljk53) [reply .jenkins/pytorch/build-mobile.sh]: > It's not _great_ (I commented on this at [#30133 (comment)](https://github.com/pytorch/pytorch/pull/30133#issuecomment-557130164)) but it's certainly the most convenient way to dump the model (and we do this in a few other places). An alternative is to save the model in s3 directly and then suck it down directly, but that makes it more difficult to update the model later.

I saw you commented there - "concerned about relying on nightlies for per PR tests" - it installs stable torch, isn't it?

File: test/mobile/custom_build/build.sh (hunk 7 above)
Comment (ljk53) [reply test/mobile/custom_build/build.sh]: removing "shows" from the comment to avoid confusing external developers :)

File: test/mobile/custom_build/prepare_model.py (hunk 3 above)
Comment (ljk53) [reply test/mobile/custom_build/prepare_model.py]: > It's copied from a very old example, so the latest approach should be `script_model = torch.jit.script(mode)`?

This doesn't work - so is there TorchScript MobileNetV2 in torchvision? or are you suggesting I should use some other TorchScript model?

File: test/mobile/custom_build/build.sh (hunk 2 above)
Comment (ezyang) [reply test/mobile/custom_build/build.sh]: "oops"

File: test/mobile/custom_build/predictor.cpp (hunk 6 above)
Comment (ezyang) [reply test/mobile/custom_build/predictor.cpp]: Ideally, it wouldn't be necessary to write this boilerplate at all! I guess it would be better to figure out who added this in the first place and ask them about it. @jerryzh168, any comments here?

File: test/mobile/custom_build/build.sh (hunk 7 above)
Comment (ezyang) [reply test/mobile/custom_build/build.sh]: Not too sure xuhdev's stuff in setup.py is all that useful, since it's specific for building PyTorch on Windows. But worth taking a look :)

File: test/mobile/custom_build/predictor.cpp (hunk 6 above)
Comment (ljk53) [reply test/mobile/custom_build/predictor.cpp]: > Ideally, it wouldn't be necessary to write this boilerplate at all! I guess it would be better to figure out who added this in the first place and ask them about it. @jerryzh168, any comments here?

I think this is introduced since @supriyar unified fbgemm & qnnpack interface at dispatcher level. Now they look the same but the underlying packed weights format / etc are different. So we introduced the concept called "qengine" to switch between these two quantization implementations. And it's possible to have both qnnpack/fbgemm available in one build so I guess it needs to be set explicitly. Yet another dispatching mechanism? :)

Comment (supriyar) [reply test/mobile/custom_build/predictor.cpp]: For mobile build though we turn off FBGEMM, right? In that case the default `qengine` should be set to QNNPACK. We control that here - https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/Context.cpp#L122

Comment (jerryzh168) [reply test/mobile/custom_build/predictor.cpp]: yeah it is another runtime dispatcher because we unified the API for fbgemm and qnnpack.

Comment (jerryzh168) [reply test/mobile/custom_build/predictor.cpp]: @supriyar did you add this boilerplate code? could you talk about why did we do this? Do we ever run this code with fbgemm backend?

Comment (supriyar) [reply test/mobile/custom_build/predictor.cpp]: @jerryzh168 see my comment above. I don't think this boilerplate is necessary because we set a default backend (for mobile it should be qnnpack) in Context.cpp.
